# Loan Approval Analytics and Prediction System

## Task 1 — Data Handling and Dataset Integration

This study examines loan application records with the objective of establishing a reliable analytical foundation for evaluating loan approval decisions. The information required for the analysis is distributed across two datasets; therefore, the initial stage focuses on understanding the structure and content of each data source before integration.

Prior to combining the datasets, their dimensions, attributes, data types and potential identifying variables are examined independently. Particular attention is given to identifying a suitable primary key and verifying its integrity through uniqueness, missing-value and duplication checks. This validation is necessary because an inappropriate integration strategy could introduce duplicate observations, unmatched applications or information loss, potentially affecting subsequent exploratory analysis and model development.

### 1.1 Importing Required Libraries

In [1]:
# Core libraries for data manipulation
import pandas as pd
import numpy as np

# Libraries for data visualisation
import matplotlib.pyplot as plt
import seaborn as sns

# Improve dataframe display within the notebook
pd.set_option("display.max_columns", None)
pd.set_option("display.max_rows", 100)

print("Required libraries imported successfully.")

Required libraries imported successfully.


### 1.2 Importing the Loan Approval Datasets

The two source datasets are imported independently using Pandas so that their structures can be examined before integration. Keeping the datasets separate at this stage allows their dimensions, attributes and record-level identifiers to be validated prior to selecting an appropriate merging strategy.

In [2]:
# Import the two source datasets
df1 = pd.read_csv("../data/Loan_Approval_Data_Set_1.csv")
df2 = pd.read_csv("../data/Loan_Approval_Data_Set_2.csv")

print("Dataset 1 loaded successfully.")
print("Dataset 2 loaded successfully.")

Dataset 1 loaded successfully.
Dataset 2 loaded successfully.


In [3]:
print(f"Dataset 1 dimensions: {df1.shape[0]:,} rows × {df1.shape[1]} columns")
print(f"Dataset 2 dimensions: {df2.shape[0]:,} rows × {df2.shape[1]} columns")

Dataset 1 dimensions: 20,000 rows × 16 columns
Dataset 2 dimensions: 20,000 rows × 18 columns


### 1.3 Initial Inspection of Dataset 1

Dataset 1 contains information relating primarily to applicants' employment, income, existing debt commitments, credit history and overall financial position. It also includes "LoanApproved", the binary response variable representing the final loan decision and therefore the principal target for the later predictive analysis.

Before integrating the two data sources, Dataset 1 is examined independently to establish its structural characteristics and identify potential data-quality concerns. The inspection considers representative observations, variable names, assigned data types and completeness. Although "ID" appears to represent an application-level identifier, its validity as a primary key will be formally assessed through uniqueness, duplication and missing-value checks before the datasets are merged.

In [4]:
# Display a sample of Dataset 1
display(df1.head())


,ID,ApplicationDate,AnnualIncome,EmploymentStatus,Experience,LoanDuration,NumberOfDependents,MonthlyDebtPayments,NumberOfOpenCreditLines,BankruptcyHistory,PreviousLoanDefaults,LengthOfCreditHistory,CheckingAccountBalance,TotalLiabilities,NetWorth,LoanApproved
0,A001,01-01-2018,39948,Employed,22,48,2,183,1,0,0,9,1202,19183,126928,0
1,A002,02-01-2018,39709,Employed,15,48,1,496,5,0,0,9,3460,9595,43609,0
2,A003,03-01-2018,40724,Employed,26,36,2,902,2,0,0,22,895,128874,5205,0
3,A004,04-01-2018,69084,Employed,34,96,1,755,2,0,0,10,1217,5370,99452,0
4,A005,05-01-2018,103264,Employed,17,36,1,274,0,0,0,27,4981,17286,227019,1


In [5]:
# Examine column names, non-null counts and data types
df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20000 entries, 0 to 19999
Data columns (total 16 columns):
 #   Column                   Non-Null Count  Dtype 
---  ------                   --------------  ----- 
 0   ID                       20000 non-null  object
 1   ApplicationDate          20000 non-null  object
 2   AnnualIncome             20000 non-null  int64 
 3   EmploymentStatus         20000 non-null  object
 4   Experience               20000 non-null  int64 
 5   LoanDuration             20000 non-null  int64 
 6   NumberOfDependents       20000 non-null  int64 
 7   MonthlyDebtPayments      20000 non-null  int64 
 8   NumberOfOpenCreditLines  20000 non-null  int64 
 9   BankruptcyHistory        20000 non-null  int64 
 10  PreviousLoanDefaults     20000 non-null  int64 
 11  LengthOfCreditHistory    20000 non-null  int64 
 12  CheckingAccountBalance   20000 non-null  int64 
 13  TotalLiabilities         20000 non-null  int64 
 14  NetWorth                 20000 non-nul

In [6]:
# Summarise the data types in Dataset 1
dataset1_types = pd.DataFrame({
    "Column": df1.columns,
    "Data_Type": df1.dtypes.astype(str).values
})

display(dataset1_types)

,Column,Data_Type
0,ID,object
1,ApplicationDate,object
2,AnnualIncome,int64
3,EmploymentStatus,object
4,Experience,int64
5,LoanDuration,int64
6,NumberOfDependents,int64
7,MonthlyDebtPayments,int64
8,NumberOfOpenCreditLines,int64
9,BankruptcyHistory,int64


#### Dataset 1 — Initial Observations

The structural inspection shows that Dataset 1 contains 20,000 loan application records and 16 attributes. Thirteen variables are currently represented as integer-valued numerical features, while "ID", "ApplicationDate", and "EmploymentStatus" have been imported as object-type variables. No null values are reported at this initial stage, indicating that Dataset 1 is structurally complete with respect to conventional missing-value representations.

Although "ApplicationDate" is currently stored as an object, it represents temporal information and should therefore be converted to an appropriate datetime format during preprocessing. Similarly, while "LoanApproved" is numerically encoded as 0 and 1, it represents a binary categorical outcome rather than a continuous numerical measurement. These distinctions will be considered during subsequent preprocessing and modelling.

The absence of conventional null values does not, however, establish that the dataset is completely free from data-quality issues. Further investigation is required to identify duplicated records, invalid values, unusual distributions and potentially inconsistent representations.

### 1.4 Initial Inspection of Dataset 2

Dataset 2 extends the applicant information available in the first data source by incorporating demographic characteristics, educational background, credit behaviour, requested loan information, asset position and risk-related indicators. These variables provide additional dimensions that may help explain differences in loan approval outcomes.

Prior to integration, the dataset is examined independently to assess its dimensions, variable types and completeness. The "ID" attribute is of particular interest because it is also present in Dataset 1 and may provide the common identifier required to associate corresponding applicant records.

In [7]:
# Display a sample of Dataset 2
display(df2.head())

,ApplicationDate,Age,CreditScore,EducationLevel,Experience,LoanAmount,MaritalStatus,HomeOwnershipStatus,ID,CreditCardUtilizationRate,NumberOfCreditInquiries,LoanPurpose,PaymentHistory,SavingsAccountBalance,TotalAssets,JobTenure,InterestRate,RiskScore
0,01-01-2018,45,617,Master,22.0,13152,Married,Own,A001,35.40,2,Home,29,7632,146111,11,22.75,49
1,02-01-2018,38,628,Associate,NaN,26045,Single,Mortgage,A002,8.78,3,Debt Consolidation,21,4627,53204,3,20.10,XX
2,03-01-2018,47,570,Bachelor,26.0,17627,Married,Rent,A003,13.70,0,Education,20,886,25176,6,21.25,XX
3,04-01-2018,58,545,High School,34.0,37898,Single,Mortgage,A004,2.67,1,Home,27,1675,104822,5,30.09,54
4,05-01-2018,37,594,Associate,17.0,9184,Married,Mortgage,A005,3.20,0,Debt Consolidation,26,1555,244305,5,17.59,36


In [8]:
# Examine column names, non-null counts and data types
df2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20000 entries, 0 to 19999
Data columns (total 18 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   ApplicationDate            20000 non-null  object 
 1   Age                        20000 non-null  int64  
 2   CreditScore                20000 non-null  int64  
 3   EducationLevel             20000 non-null  object 
 4   Experience                 19334 non-null  float64
 5   LoanAmount                 20000 non-null  int64  
 6   MaritalStatus              19985 non-null  object 
 7   HomeOwnershipStatus        20000 non-null  object 
 8   ID                         20000 non-null  object 
 9   CreditCardUtilizationRate  20000 non-null  float64
 10  NumberOfCreditInquiries    20000 non-null  int64  
 11  LoanPurpose                20000 non-null  object 
 12  PaymentHistory             20000 non-null  int64  
 13  SavingsAccountBalance      20000 non-null  int

In [9]:
# Summarise the data types in Dataset 2
dataset2_types = pd.DataFrame({
    "Column": df2.columns,
    "Data_Type": df2.dtypes.astype(str).values
})

display(dataset2_types)

,Column,Data_Type
0,ApplicationDate,object
1,Age,int64
2,CreditScore,int64
3,EducationLevel,object
4,Experience,float64
5,LoanAmount,int64
6,MaritalStatus,object
7,HomeOwnershipStatus,object
8,ID,object
9,CreditCardUtilizationRate,float64


#### Dataset 2 — Initial Observations

Dataset 2 contains 20,000 observations and 18 attributes, comprising eight integer, three floating-point and seven object-type variables. Unlike Dataset 1, the initial structural assessment identifies explicit missing observations in two attributes. "Experience" contains 19,334 non-null records, indicating 666 missing values, while "MaritalStatus" contains 19,985 non-null records, corresponding to 15 missing observations.

A further data-quality concern is evident in "RiskScore". Although Pandas reports 20,000 non-null observations for this variable, it has been imported as an object rather than a numerical data type. Inspection of the source records indicates that non-numeric placeholders such as "XX" occur within this field. Consequently, the non-null count alone understates the true extent of missing or invalid information because these placeholders are interpreted as text rather than conventional null values.

"ApplicationDate" has similarly been imported as an object and will require conversion to a datetime representation before time-based analysis. These findings demonstrate the importance of examining both data types and underlying values rather than relying exclusively on conventional null-value counts when assessing data quality.

### 1.5 Primary Key Validation

The "ID" attribute occurs in both source datasets and therefore represents the principal candidate for linking corresponding loan application records. However, the presence of a common field does not by itself establish that it is a valid primary key. Prior to integration, the identifier is evaluated for completeness and uniqueness within each dataset, followed by an assessment of the degree of correspondence between the two sets of identifiers.

A suitable application-level key should contain no missing values, should uniquely identify individual records within each dataset, and should provide consistent correspondence between records across the two data sources. These checks reduce the risk of creating duplicate observations or unintended many-to-many relationships during the merge operation.

In [10]:
# Validate ID as a candidate primary key in both datasets
key_validation = pd.DataFrame({
    "Dataset": ["Dataset 1", "Dataset 2"],
    "Total_Rows": [len(df1), len(df2)],
    "Unique_IDs": [
        df1["ID"].nunique(),
        df2["ID"].nunique()
    ],
    "Missing_IDs": [
        df1["ID"].isna().sum(),
        df2["ID"].isna().sum()
    ],
    "Duplicate_IDs": [
        df1["ID"].duplicated().sum(),
        df2["ID"].duplicated().sum()
    ]
})

display(key_validation)

,Dataset,Total_Rows,Unique_IDs,Missing_IDs,Duplicate_IDs
0,Dataset 1,20000,20000,0,0
1,Dataset 2,20000,20000,0,0


In [11]:
# Examine the correspondence of IDs across the two datasets
ids_df1 = set(df1["ID"].dropna())
ids_df2 = set(df2["ID"].dropna())

common_ids = ids_df1.intersection(ids_df2)
only_df1 = ids_df1.difference(ids_df2)
only_df2 = ids_df2.difference(ids_df1)

print(f"Unique IDs in Dataset 1 : {len(ids_df1):,}")
print(f"Unique IDs in Dataset 2 : {len(ids_df2):,}")
print(f"IDs common to both       : {len(common_ids):,}")
print(f"IDs only in Dataset 1    : {len(only_df1):,}")
print(f"IDs only in Dataset 2    : {len(only_df2):,}")

Unique IDs in Dataset 1 : 20,000
Unique IDs in Dataset 2 : 20,000
IDs common to both       : 20,000
IDs only in Dataset 1    : 0
IDs only in Dataset 2    : 0


In [12]:
# Identify variables occurring in both source datasets
common_columns = sorted(
    set(df1.columns).intersection(df2.columns)
)

print("Columns shared by both datasets:")
for column in common_columns:
    print(f"- {column}")

Columns shared by both datasets:
- ApplicationDate
- Experience
- ID


#### Primary Key Assessment

The validation results provide strong evidence that "ID" is an appropriate primary key for integrating the two data sources. Each dataset contains 20,000 records and exactly 20,000 unique identifiers, with no missing or duplicated "ID" values. This establishes uniqueness and completeness of the candidate key within each individual dataset.

Cross-dataset comparison further shows that all 20,000 identifiers are shared between the two datasets, with no application identifiers occurring exclusively in either source. The datasets therefore exhibit complete one-to-one correspondence at the application level. On this basis, "ID" can be used as the primary merge key while preserving the expected 20,000 application records.

The schema comparison also identifies "ApplicationDate" and "Experience" as attributes occurring in both datasets. These overlapping variables require additional consistency checks before integration to avoid retaining redundant or conflicting versions of the same information.

In [13]:
# Compare ApplicationDate values between the two datasets by ID
date_comparison = df1[["ID", "ApplicationDate"]].merge(
    df2[["ID", "ApplicationDate"]],
    on="ID",
    how="inner",
    suffixes=("_df1", "_df2"),
    validate="one_to_one"
)

date_comparison["Date_Match"] = (
    date_comparison["ApplicationDate_df1"]
    == date_comparison["ApplicationDate_df2"]
)

print("ApplicationDate consistency check")
print("-" * 40)
print(f"Matching dates    : {date_comparison['Date_Match'].sum():,}")
print(f"Non-matching dates: {(~date_comparison['Date_Match']).sum():,}")

ApplicationDate consistency check
----------------------------------------
Matching dates    : 2,922
Non-matching dates: 17,078


In [14]:
# Compare Experience values between datasets using the common ID
experience_comparison = df1[["ID", "Experience"]].merge(
    df2[["ID", "Experience"]],
    on="ID",
    how="inner",
    suffixes=("_df1", "_df2"),
    validate="one_to_one"
)

# Compare values only where Dataset 2 contains Experience
available_experience = experience_comparison[
    experience_comparison["Experience_df2"].notna()
].copy()

experience_matches = (
    available_experience["Experience_df1"]
    == available_experience["Experience_df2"]
)

print("Experience consistency check")
print("-" * 40)
print(f"Dataset 1 non-null Experience : {df1['Experience'].notna().sum():,}")
print(f"Dataset 2 non-null Experience : {df2['Experience'].notna().sum():,}")
print(f"Dataset 2 missing Experience  : {df2['Experience'].isna().sum():,}")
print(f"Comparable records            : {len(available_experience):,}")
print(f"Matching values               : {experience_matches.sum():,}")
print(f"Non-matching values           : {(~experience_matches).sum():,}")


Experience consistency check
----------------------------------------
Dataset 1 non-null Experience : 20,000
Dataset 2 non-null Experience : 19,334
Dataset 2 missing Experience  : 666
Comparable records            : 19,334
Matching values               : 19,334
Non-matching values           : 0


### 1.6 Investigation of Overlapping Attributes

The comparison of the duplicated "Experience" attribute demonstrates complete agreement across all 19,334 records for which Dataset 2 contains an observed value. Dataset 2 nevertheless contains 666 missing "Experience" observations, whereas Dataset 1 provides a complete value for every applicant. This suggests that the version contained in Dataset 1 can be retained during integration without requiring statistical imputation, subject to the established cross-source consistency.

In contrast, direct comparison of "ApplicationDate" produces a substantial number of apparent mismatches. Since the variable was imported as an object in both datasets, these differences may reflect inconsistent textual date representations rather than genuine disagreement in application dates. The discrepant records are therefore examined before determining how the duplicated date fields should be reconciled.

In [15]:
# Inspect examples where ApplicationDate does not match
date_mismatches = date_comparison[
    ~date_comparison["Date_Match"]
].copy()

display(date_mismatches.head(15))

,ID,ApplicationDate_df1,ApplicationDate_df2,Date_Match
2922,A2923,01-01-2018,01-01-2026,False
2923,A2924,02-01-2018,02-01-2026,False
2924,A2925,03-01-2018,03-01-2026,False
2925,A2926,04-01-2018,04-01-2026,False
2926,A2927,05-01-2018,05-01-2026,False
2927,A2928,06-01-2018,06-01-2026,False
2928,A2929,07-01-2018,07-01-2026,False
2929,A2930,08-01-2018,08-01-2026,False
2930,A2931,09-01-2018,09-01-2026,False
2931,A2932,10-01-2018,10-01-2026,False


In [16]:
# Inspect the original date formats in both datasets
print("Dataset 1 date examples:")
print(df1["ApplicationDate"].head(15).tolist())

print("\nDataset 2 date examples:")
print(df2["ApplicationDate"].head(15).tolist())

Dataset 1 date examples:
['01-01-2018', '02-01-2018', '03-01-2018', '04-01-2018', '05-01-2018', '06-01-2018', '07-01-2018', '08-01-2018', '09-01-2018', '10-01-2018', '11-01-2018', '12-01-2018', '13-01-2018', '14-01-2018', '15-01-2018']

Dataset 2 date examples:
['01-01-2018', '02-01-2018', '03-01-2018', '04-01-2018', '05-01-2018', '06-01-2018', '07-01-2018', '08-01-2018', '09-01-2018', '10-01-2018', '11-01-2018', '12-01-2018', '13-01-2018', '14-01-2018', '15-01-2018']


In [17]:
# Convert ApplicationDate to datetime using the known day-month-year format
df1_dates = pd.to_datetime(
    df1["ApplicationDate"],
    format="%d-%m-%Y",
    errors="coerce"
)

df2_dates = pd.to_datetime(
    df2["ApplicationDate"],
    format="%d-%m-%Y",
    errors="coerce"
)

print("Dataset 1")
print(f"Earliest date : {df1_dates.min().date()}")
print(f"Latest date   : {df1_dates.max().date()}")
print(f"Invalid dates : {df1_dates.isna().sum():,}")

print("\nDataset 2")
print(f"Earliest date : {df2_dates.min().date()}")
print(f"Latest date   : {df2_dates.max().date()}")
print(f"Invalid dates : {df2_dates.isna().sum():,}")

Dataset 1
Earliest date : 2018-01-01
Latest date   : 2025-12-31
Invalid dates : 0

Dataset 2
Earliest date : 2018-01-01
Latest date   : 2072-10-03
Invalid dates : 0


In [18]:
# Examine the distribution of records by year
year_comparison = pd.DataFrame({
    "Dataset_1": df1_dates.dt.year.value_counts().sort_index(),
    "Dataset_2": df2_dates.dt.year.value_counts().sort_index()
}).fillna(0).astype(int)

display(year_comparison)

,Dataset_1,Dataset_2
ApplicationDate,,
2018,2555,365
2019,2555,365
2020,2562,366
2021,2555,365
2022,2555,365
2023,2555,365
2024,2473,366
2025,2190,365
2026,0,365


In [19]:
# Examine agreement between the two date fields after datetime conversion
date_check = pd.DataFrame({
    "ID": df1["ID"],
    "Date_df1": df1_dates,
    "Date_df2": df2_dates
})

date_check["Match"] = date_check["Date_df1"] == date_check["Date_df2"]

print(f"Matching records     : {date_check['Match'].sum():,}")
print(f"Non-matching records : {(~date_check['Match']).sum():,}")

print("\nFirst five mismatches:")
display(date_check.loc[~date_check["Match"]].head())

print("Last five mismatches:")
display(date_check.loc[~date_check["Match"]].tail())

Matching records     : 2,922
Non-matching records : 17,078

First five mismatches:


,ID,Date_df1,Date_df2,Match
2922,A2923,2018-01-01,2026-01-01,False
2923,A2924,2018-01-02,2026-01-02,False
2924,A2925,2018-01-03,2026-01-03,False
2925,A2926,2018-01-04,2026-01-04,False
2926,A2927,2018-01-05,2026-01-05,False


Last five mismatches:


,ID,Date_df1,Date_df2,Match
19995,B9997,2024-09-29,2072-09-29,False
19996,B9998,2024-09-30,2072-09-30,False
19997,B9999,2024-10-01,2072-10-01,False
19998,C001,2024-10-02,2072-10-02,False
19999,C002,2024-10-03,2072-10-03,False


#### Reconciliation of Overlapping Attributes

Further examination of the duplicated attributes reveals different data-quality characteristics between the two sources. For "Experience", all 19,334 observations available in both datasets agree exactly, while Dataset 1 additionally contains valid values for the 666 observations that are missing from Dataset 2. Consequently, Dataset 1 provides the more complete version of this attribute and is retained for the integrated dataset. This cross-source reconciliation is preferable to statistical imputation because observed applicant-level information is available from the corresponding record in the alternative source.

A more substantial inconsistency is identified in "ApplicationDate". Only 2,922 of the 20,000 paired records contain identical dates. Dataset 1 covers the period from 1 January 2018 to 31 December 2025, corresponding to the stated 96-month data-collection period. In contrast, Dataset 2 extends from 2018 to 2072, with later records containing implausible future dates. The discrepancy therefore cannot reasonably be attributed solely to formatting differences.

Dataset 1 is consequently treated as the authoritative source for "ApplicationDate", while the duplicate date field from Dataset 2 is excluded during integration. This decision preserves consistency with the documented observation period and prevents erroneous temporal information from affecting subsequent trend analysis or predictive modelling. In a production environment, such a discrepancy should additionally be referred to the data owner and documented through a formal data-quality process rather than silently corrected.

### 1.7 Dataset Integration

Following validation of the primary key and reconciliation of the overlapping attributes, the two datasets are integrated at application level using "ID". Since "ApplicationDate" and "Experience" from Dataset 1 have been selected as the authoritative versions, their duplicate counterparts are removed from Dataset 2 before merging.

A one-to-one merge constraint is applied explicitly. This provides an additional integrity safeguard by causing the operation to fail if either source unexpectedly contains multiple records for the same identifier.

In [20]:
# Remove duplicate attributes from Dataset 2 before integration
df2_for_merge = df2.drop(
    columns=["ApplicationDate", "Experience"]
).copy()

# Merge the datasets using the validated primary key
loan_data = df1.merge(
    df2_for_merge,
    on="ID",
    how="inner",
    validate="one_to_one"
)

print("Datasets merged successfully.")
print(f"Rows after merge    : {loan_data.shape[0]:,}")
print(f"Columns after merge : {loan_data.shape[1]}")

Datasets merged successfully.
Rows after merge    : 20,000
Columns after merge : 31


In [21]:
# Post-merge integrity checks
print("POST-MERGE VALIDATION")
print("-" * 45)

print(f"Expected records       : {len(df1):,}")
print(f"Actual merged records  : {len(loan_data):,}")
print(f"Unique IDs             : {loan_data['ID'].nunique():,}")
print(f"Missing IDs            : {loan_data['ID'].isna().sum():,}")
print(f"Duplicate IDs          : {loan_data['ID'].duplicated().sum():,}")

print(
    "All Dataset 1 IDs retained:",
    set(df1["ID"]) == set(loan_data["ID"])
)

print(
    "All Dataset 2 IDs retained:",
    set(df2["ID"]) == set(loan_data["ID"])
)

POST-MERGE VALIDATION
---------------------------------------------
Expected records       : 20,000
Actual merged records  : 20,000
Unique IDs             : 20,000
Missing IDs            : 0
Duplicate IDs          : 0
All Dataset 1 IDs retained: True
All Dataset 2 IDs retained: True


In [22]:
# Inspect the integrated analytical dataset
display(loan_data.head())

print(f"\nFinal Task 1 dataset shape: {loan_data.shape}")

,ID,ApplicationDate,AnnualIncome,EmploymentStatus,Experience,LoanDuration,NumberOfDependents,MonthlyDebtPayments,NumberOfOpenCreditLines,BankruptcyHistory,PreviousLoanDefaults,LengthOfCreditHistory,CheckingAccountBalance,TotalLiabilities,NetWorth,LoanApproved,Age,CreditScore,EducationLevel,LoanAmount,MaritalStatus,HomeOwnershipStatus,CreditCardUtilizationRate,NumberOfCreditInquiries,LoanPurpose,PaymentHistory,SavingsAccountBalance,TotalAssets,JobTenure,InterestRate,RiskScore
0,A001,01-01-2018,39948,Employed,22,48,2,183,1,0,0,9,1202,19183,126928,0,45,617,Master,13152,Married,Own,35.40,2,Home,29,7632,146111,11,22.75,49
1,A002,02-01-2018,39709,Employed,15,48,1,496,5,0,0,9,3460,9595,43609,0,38,628,Associate,26045,Single,Mortgage,8.78,3,Debt Consolidation,21,4627,53204,3,20.10,XX
2,A003,03-01-2018,40724,Employed,26,36,2,902,2,0,0,22,895,128874,5205,0,47,570,Bachelor,17627,Married,Rent,13.70,0,Education,20,886,25176,6,21.25,XX
3,A004,04-01-2018,69084,Employed,34,96,1,755,2,0,0,10,1217,5370,99452,0,58,545,High School,37898,Single,Mortgage,2.67,1,Home,27,1675,104822,5,30.09,54
4,A005,05-01-2018,103264,Employed,17,36,1,274,0,0,0,27,4981,17286,227019,1,37,594,Associate,9184,Married,Mortgage,3.20,0,Debt Consolidation,26,1555,244305,5,17.59,36



Final Task 1 dataset shape: (20000, 31)


### 1.8 Task 1 Summary

The data integration stage established a reliable analytical foundation by validating the structure and compatibility of the two source datasets before combining them. Both sources contained 20,000 application records, and "ID" was confirmed as a suitable primary key through tests of uniqueness, completeness and cross-dataset correspondence. No missing or duplicated identifiers were identified, and every application ID was represented in both datasets.

The assessment of overlapping attributes also identified important data-quality differences between the sources. "Experience" showed complete agreement across all records where values were available in both datasets; however, Dataset 2 contained 666 missing observations. Since Dataset 1 contained a complete and consistent version of this attribute, it was retained rather than introducing unnecessary statistical imputation. The duplicated "ApplicationDate" fields presented a more significant inconsistency. Dataset 1 covered January 2018 to December 2025, which corresponds to the documented 96-month observation period, whereas Dataset 2 extended to 2072. Dataset 1 was therefore selected as the authoritative source for the application date.

Following these validation and reconciliation procedures, the datasets were integrated through a one-to-one inner join on "I"`. The resulting analytical dataset contains 20,000 observations and 31 attributes, with no loss of applicant records, missing identifiers or duplicated identifiers. This integrated dataset will form the basis of the subsequent exploratory analysis, preprocessing, feature engineering and predictive modelling stages.